### DISCUSSION TITLE
Filter single events

### DISCUSSION THREAD

**Discussion by dachickenuser** [derekantrican/GAS-ICS-Sync#376] [Q&A] [answered]
What would be the most appropriate way to filter single events by event name/title and/or calendar name?

**Comment by jonas0b1011001** [comment, marked as answer]
Have a look if [this branch](https://github.com/derekantrican/GAS-ICS-Sync/tree/event-filtering) suits your needs.

**Reply by rtromao** [reply]
Perhaps I'm doing something wrong. I have the same question (filter events with a specific title) and following the link in the previous post does not take me anywhere. Could someone help me?

**Reply by jonas0b1011001** [reply]
The link should lead to a different branch of the repo. Follow [https://github.com/derekantrican/GAS-ICS-Sync/wiki/Setting-up-the-script-manually](https://github.com/derekantrican/GAS-ICS-Sync/wiki/Setting-up-the-script-manually) to set up all *.gs files manually.

To filter specific event titles, add a filter in filters.gs:
```
{ 
  parameter: "summary",
  type: "exclude", //include/exclude
  comparison: "regex", // equals, begins with, contains, regex
  criterias: ["xyz"]
}
```

**Reply by rtromao** [reply]
ohh... now I got your point. What I should do is to create another project in Google Apps Script using this branch and customizing the filter as needed. Is that correct? Thank you so much for your support 😊

**Reply by jonas0b1011001** [reply]
Correct. Let me know if you get stuck again.

**Reply by rtromao** [reply]
It is a shame, but I'm stuck 😒. I did copy the project to my GAS page, added a filters.gs file with the same format found in Github, modified the parameter summary to type: "exclude", comparison: "equals" and criterias: ["Away"].
After syncing I still found new entries with the title Away. 
Perhaps I'm wrong, but I can't see any reference in the code.gs to the filters var. Is something missing?
Thanks and sorry for bothering.

**Reply by jonas0b1011001** [reply]
[Helpers.gs](https://github.com/derekantrican/GAS-ICS-Sync/blob/event-filtering/Helpers.gs) is different from master branch, make sure you got the correcrt version.

https://github.com/derekantrican/GAS-ICS-Sync/blob/430f83d3674f2582eb664e35fa353084dec2c853/Helpers.gs#L214

**Reply by rtromao** [reply]
Well... i was using the incorrect helper but the behavior is the same as before even atter updating the file. I believe now everything is the way it is supposed to be, but still no working....

**Reply by jonas0b1011001** [reply]
Can you share your filters.gs?

**Reply by rtromao** [reply]
Sure...

var filters = [
            /*{ 
              parameter: "summary",
              type: "exclude", //include/exclude
              comparison: "equals", // equals, begins with, contains, regex
              criterias: ["Away"]
            },
            {
              parameter: "categories",
              type: "include", //include/exclude
              comparison: "equals", // equals, begins with, contains, regex
              criterias: ["Meetings"]
            },
            {
              parameter: "dtend",
              type: "include", //include/exclude
              comparison: ">", // <,>, =
              offset: -7
            },
            {
              parameter: "dtstart",
              type: "exclude", //include/exclude
              comparison: ">", // <,>, =
              offset: 14
            } */
          ];

**Reply by jonas0b1011001** [reply]
The filters are commented out by default as they are just examples.
Change it to

```
var filters = [
{
parameter: "summary",
type: "exclude", //include/exclude
comparison: "equals", // equals, begins with, contains, regex
criterias: ["Away"]
}
];
```

and you should be good to go.

**Reply by rtromao** [reply]
I didn't realize the asterisk in the code... sorry. Anyway, still not working... Please see the Outlook and Google screenshots attached.
Please, keep in mind Away and Busy are not the real event details, this is what I can craw from the shared Outlook agenda, not sure if this matter or not..... 
![2024-01-26_160430](https://github.com/derekantrican/GAS-ICS-Sync/assets/40307398/6c78ff16-5436-4f01-99d7-cd7fb54adc75)

**Reply by jonas0b1011001** [reply]
> Please, keep in mind Away and Busy are not the real event details, this is what I can craw from the shared Outlook agenda, not sure if this matter or not.....

That should not make a difference. Have you uninstalled your previous version of the script, maybe it is still adding the events?

Try making a copy of [this script](https://script.google.com/d/1tEl8TmHUNY5hZqFSm5FsqosfiDulMTqrXbNJCnBAmKvqtKOxDTv41RJX/edit?usp=sharing), it's the same branch fully set up. You just have to add your source calendar.

**Reply by rtromao** [reply]
This one worked like a charm. No more Away events in my Google Calendar. Thank you so much for your help. Appreciate 😊😊😊

**Comment by javier-molina** [comment]
This is a welcome addition but I'm haven't been able to match an event for exclusion:

I have merged this branch into a copy of master locally.

I have configured this rule below

``` 
{ 
              parameter:"SUMMARY",
              type: "exclude", //include/exclude
              comparison: "contains", // equals, begins with, contains, regex
              criterias: ["Working elsewhere"]
            
}

```
But it seems the event below is never matched and hence still added to google calendar

```
BEGIN:VEVENT
UID:040000008200E00074C5B7101A82E008000000004299D4CE6791DA01000000000000000
 010000000BC1C1FFA4B687249B4504C5A73C51098
SUMMARY:Working elsewhere
DTSTART;TZID=AUS Eastern Standard Time:20240422T090000
DTEND;TZID=AUS Eastern Standard Time:20240422T093000
CLASS:PUBLIC
PRIORITY:5
DTSTAMP:20240418T082002Z
TRANSP:OPAQUE
STATUS:CONFIRMED
SEQUENCE:0
X-MICROSOFT-CDO-APPT-SEQUENCE:0
X-MICROSOFT-CDO-BUSYSTATUS:WORKINGELSEWHERE
X-MICROSOFT-CDO-INTENDEDSTATUS:BUSY
X-MICROSOFT-CDO-ALLDAYEVENT:FALSE
X-MICROSOFT-CDO-IMPORTANCE:1
X-MICROSOFT-CDO-INSTTYPE:0
X-MICROSOFT-DONOTFORWARDMEETING:FALSE
X-MICROSOFT-DISALLOW-COUNTER:FALSE
X-MICROSOFT-REQUESTEDATTENDANCEMODE:DEFAULT
X-MICROSOFT-ISRESPONSEREQUESTED:FALSE
END:VEVENT
```

Here is the output when running (I have 5 events in my test calendar)

```
7:38:04 PM	Notice	Execution started
7:38:06 PM	Info	Syncing 1 calendars to Work busy calendar
7:38:06 PM	Info	Working on calendar: 10d81e7987430b00ef6d7e0791c85040782fcb8d275349768b34a6676c56c764@group.calendar.google.com
7:38:06 PM	Info	Fetched 4 existing events from Work busy calendar
7:38:06 PM	Info	Applying 1 filters on 5 events.
7:38:06 PM	Info	5 events left.
7:38:06 PM	Info	Parsed 5 events from ical sources
7:38:06 PM	Info	Processing 5 events
7:38:06 PM	Info	Adjusted RRule/RDate to exclude past instances
7:38:06 PM	Info	Skipping unchanged Event 040000008200E00074C5B7101A82E00800000000152CA7C22F54DA010000000000000000100000008896FC0FA8A86A46823A78416F33E64C
7:38:06 PM	Info	Skipping unchanged Event 040000008200E00074C5B7101A82E008000000006153A9C16791DA010000000000000000100000002D7B838FDE38FE42A7940D0FEE484D5C
7:38:06 PM	Info	Converting ICS timezone AUS Eastern Standard Time to Google Calendar (IANA) timezone Australia/Sydney
7:38:06 PM	Info	Converting ICS timezone AUS Eastern Standard Time to Google Calendar (IANA) timezone Australia/Sydney
7:38:06 PM	Info	Adding new event 040000008200E00074C5B7101A82E008000000004299D4CE6791DA01000000000000000010000000BC1C1FFA4B687249B4504C5A73C51098
7:38:07 PM	Info	Skipping unchanged Event 040000008200E00074C5B7101A82E00800000000FC71F5DC6791DA01000000000000000010000000D3B2AF52D257D14689099B260D9B4269
7:38:07 PM	Info	Skipping unchanged Event 040000008200E00074C5B7101A82E00800000000F046AEE96791DA010000000000000000100000002C79493076F11E4FBA01908EF31FBC8D
7:38:07 PM	Info	Done processing events
7:38:07 PM	Info	Checking 4 events for removal
7:38:07 PM	Info	Done checking events for removal
7:38:07 PM	Info	Processing 0 Recurrence Instances!
7:38:07 PM	Info	Sync finished!
7:38:08 PM	Notice	Execution completed


```

The Checking 4 events for removal is because I had manually removed one entry in my google calendar from a previous run.

This filter does not work either

```
{ 
              parameter:"X-MICROSOFT-CDO-BUSYSTATUS",
              type: "exclude", //include/exclude
              comparison: "equals", // equals, begins with, contains, regex
              criterias: ["WORKINGELSEWHERE"]
            }
``` 

What am I doing wrong?

Thanks
J

**Reply by jonas0b1011001** [reply]
use all lower case for the parameter setting.
`parameter: "summary",`

Will add a fix fo that soon.

**Reply by javier-molina** [reply]
Thanks, that fixed the issue.

Happy filtering now!
